game/qsp.ksy

```diff
@@ -0,0 +1,353 @@
+        doc: |
+          import struct
+          sizes = {1: "B", 2: "H", 4: "I", 8: "Q"}
+          endiannesses = {"": "", "le": "<", "be": ">"}
+          
+          def isPowerOf2(x: int) -> bool:
+              return bool((x - 1) & x)
+          
+          class GFOp:
+              __slots__ = ("chunkSize", "s", "scaleFixer", "min")
+              
+              def __init__(self, chunkSize: int = 1, endianness: str = "", min: int = 0, scale: int = None):
+                  self.chunkSize = chunkSize
+                  self.s = struct.Struct("".join((endiannesses[endianness], sizes[chunkSize])))
+                  if scale is None:
+                      scale = 2 ** (8 * chunkSize) - min
+                  
+                  if isPowerOf2(scale):
+                      mask = scale - 1
+                      self.scaleFixer = lambda x: x & mask
+                  else:
+                      self.scaleFixer = lambda x: x % scale
+                  
+                  self.min = min
+              
+              def processFunc(self, val):
+                  raise NotImplementedError()
+              
+              def decode(self, data):
+                  l = len(data)
+                  result = bytearray(l)
+                  ptr = 0
+                  while ptr < l:
+                      nextPtr = ptr + self.chunkSize
+                      result[ptr:nextPtr] = self.s.pack(self.processFunc(self.s.unpack(data[ptr:nextPtr])[0]))
+                      ptr = nextPtr
+                  return bytes(result)
+          
+          class Rot(GFOp):
+              __slots__ = ("rotValue",)
+              
+              def __init__(self, rotValue: int = 13, chunkSize: int = 1, endianness: str = "", min: int = 0, scale: int = None):
+                  super().__init__(chunkSize=chunkSize, endianness=endianness, min=min, scale=scale)
+                  self.rotValue = rotValue
+              
+              def processFunc(self, val):
+                  return self.min + self.scaleFixer(val - self.min + self.rotValue)
+          
+          r = Rot()
+          r.decode(r.decode(b"abcd"))
+
+  number_str:
+    -orig-id: "#число" # 'number' in Russian
+    seq:
+      - id: str
+        type: usual_str
+    instances:
+      number:
+        value: str.val.to_i
+  rot5_number_str:
+    -orig-id: "##кодированное число" # 'encoded number' in Russian
+    seq:
+      - id: str
+        type: rot5_str
+    instances:
+      number:
+        value: str.val.to_i
+  loc:
+    seq:
+      - id: name
+        type: rot5_str
+      - id: descr
+        type: rot5_str
+      - id: code
+        type: rot5_str
+      - id: actions_count_new
+        type: rot5_number_str
+        if: _root.is_new_format
+      - id: actions
+        type: act
+        repeat: expr
+        repeat-expr: actions_count
+    instances:
+      actions_count:
+        value: "(_root.is_new_format?actions_count_new.as<rot5_number_str>.number:20)"
```

Comment:
Unnecessary casting, the field `actions_count_new` is always of the type `rot5_number_str `.

Reply:
Heh, I'd love our compiler some day being able to issue warnings like that ;) Thanks for the catch!

Reply:
Thanks, fixed.